docs/howtos/exposing-rust-components-to-kotlin.md

```diff
@@ -0,0 +1,203 @@
+This code should use [JNA](https://github.com/java-native-access/jna)
+to load the C FFI via shared library
+and expose it
+as a nice safe ergonomic Kotlin API.
+
+It seems likely
+that we could provide
+a useful template
+and/or suite of shared utilities
+to make the Kotlin easier to write.
+But we haven't yet.
+
+Finally, you will need to
+add your package
+into the [android-components](https://github.com/mozilla-mobile/android-components) repo
+via some undocumented process
+that starts by asking
+in the rust-components channel on slack.
+
+### How should I name the resulting package?
+
+Great question!
+We should write
+or link to
+an answer here.
+
+### How do I publish the resulting package?
+
+Great question!
+We should write
+or link to
+an answer here.
+
+### How do I know what library name to load to access my C FFI?
+
+Great question!
+We should write
+or link to
+an answer here.
+
+### Why can’t we use cbindgen or an alternative C binding generator?
+
+We could,
+but it wouldn’t save us that much.
+cbindgen would automate
+the process of turning the ffi crate's source
+into the equivalent C header file,
+however it wouldn’t help
+with the kotlin code
+(although one could imagine something
+that automated generation
+of the kotlin bindings file).
+In particular,
+it wouldn’t help us produce
+the ffi crate’s code,
+or the wrappers that make the bindings safe
+(this is most of the work).
+It would be nice to automate the process,
+but it hasn’t been particularly error prone
+to update it by hand so far.
+
+It’s also worth noting
+that the ffi crate’s source
+doesn’t have the ownership information
+encoded in it with regards to strings
+(they’re all pointers to c_char),
+which makes generating kotlin bindings harder,
+since the kotlin bindings
+need to take/return Pointer for strings owned by Rust,
+and String for strings owned by Kotlin.
+This is because
+when we later release the string owned by rust,
+it must be the same pointer we were given earlier
+(If you pass String,
+JNA allocates temporary memory for the call,
+passes you it,
+and releases it afterwards).
+
+(The solution to this
+would likely be something like wasm_bindgen,
+where you annotate your Rust source.
+The tool would then spit out both the FFI crate,
+and the bulk of the Kotlin/Swift APIs.
+This would be a lot of work,
+but it would be cool to have someday).
+
+cbindgen doesn’t always work seamlessly.
+It’s a standalone parser of rust code,
+not a rustc plugin --
+it doesn’t always understand
+your rust library 
+like the compiler does.
+A number of these issues
+cropped up when autopush used it:
+e.g. it being unable to parse newer rust code
+or having buggy handling of certain rust syntax.
+
+### What design principles should inform my component API?
+
+Many, most of which aren't written down yet.
+This is an incomplete list:
+
+* Avoid callbacks where possible,
+  in favour of simple blocking calls.
+* Think of building a "library", not a "framework";
```

Comment:
+1! Thanks for writing this out, I'd have forgotten it, but not taking control is a big design consideration for all of our components